### ISSUE TITLE
Implement LSTM Training Pipeline with Advanced Optimization

### ISSUE THREAD

**Issue by dhar174** [dhar174/steel_defect_demo#70]
# LSTM Training Pipeline Implementation for Steel Casting Defect Prediction

## Overview
Implement a comprehensive LSTM training pipeline that extends the existing `src/models/model_trainer.py` with deep learning-specific functionality. This implementation will provide a robust, production-ready training framework with advanced optimization techniques, early stopping, comprehensive metrics tracking, and support for GPU acceleration.

### Primary Implementation
- **File**: `src/models/model_trainer.py` (extended)
- **Dependencies**: `torch>=1.13.0`, `numpy>=1.23.0`, `pyyaml>=6.0`, `tqdm>=4.64.0`, `tensorboard>=2.11.0`
- **Integration**: Direct integration with LSTM model architecture and sequence dataset components

#### 1. LSTMTrainer Class Implementation
```python
class LSTMTrainer:
    """Advanced training pipeline for LSTM model with comprehensive features"""
    
    def __init__(self, model: nn.Module, config: Dict, device: str = 'auto'):
        """
        Initialize LSTM trainer with model and configuration
        
        Parameters:
        - model: SteelDefectLSTM model instance
        - config: Training configuration from model_config.yaml
        - device: Training device ('auto', 'cpu', 'cuda')
        """
```

**Key Initialization Features:**
- Automatic device detection and GPU utilization
- Model parameter counting and architecture summary
- Configuration validation and parameter verification
- Training state initialization and checkpoint management
- Logging system setup with file and console handlers

#### 2. Adam Optimizer with Advanced Configuration
```python
def setup_optimizer(self) -> None:
    """
    Configure Adam optimizer with advanced parameter settings
    
    Features:
    - Configurable learning rate and weight decay
    - Parameter group separation for different layer types
    - Gradient clipping preparation
    - Learning rate warmup support
    """
```

**Optimizer Configuration Requirements:**
- Learning rate: 0.001 (default, configurable)
- Weight decay: 1e-4 for regularization
- Beta parameters: (0.9, 0.999) for momentum
- Epsilon: 1e-8 for numerical stability
- Support for parameter-specific learning rates
- Gradient clipping threshold: 1.0 (configurable)

#### 3. Weighted Binary Cross-Entropy Loss Implementation
```python
def setup_loss_function(self) -> None:
    """
    Configure weighted loss function for class imbalance handling
    
    Features:
    - Automatic positive weight calculation from training data
    - Configurable weight multiplier (default: 3.0)
    - Label smoothing support
    - Focal loss option for extreme imbalance
    """
```

**Loss Function Specifications:**
- Primary: `BCEWithLogitsLoss` with positive weight
- Positive weight multiplier: 3.0 (configurable)
- Automatic weight calculation based on class distribution
- Support for label smoothing (alpha = 0.1)
- Optional focal loss implementation for severe imbalance

#### 4. Early Stopping with Patience-Based Monitoring
```python
class EarlyStopping:
    """Early stopping implementation with comprehensive monitoring"""
    
    def __init__(self, patience: int = 15, min_delta: float = 1e-4, 
                 restore_best_weights: bool = True):
        """
        Initialize early stopping with validation monitoring
        
        Parameters:
        - patience: Number of epochs to wait for improvement
        - min_delta: Minimum change to qualify as improvement
        - restore_best_weights: Restore best model weights on stop
        """
```

**Early Stopping Features:**
- Validation loss monitoring with configurable patience (default: 15 epochs)
- Minimum delta threshold for improvement detection
- Best model weights restoration
- Multiple metric monitoring (loss, AUC-ROC, AUC-PR)
- Plateau detection and learning rate reduction triggers
- Training curve smoothing for robust decision making

#### 5. Learning Rate Scheduling and Optimization
```python
def setup_scheduler(self) -> None:
    """
    Configure learning rate scheduling strategy
    
    Supported schedulers:
    - ReduceLROnPlateau: Reduce LR when validation plateaus
    - CosineAnnealingLR: Cosine annealing with warm restarts
    - StepLR: Step-based learning rate decay
    - ExponentialLR: Exponential decay
    """
```

**Learning Rate Scheduling Options:**
- **Primary**: `ReduceLROnPlateau` with validation loss monitoring
- Factor: 0.5 (halve learning rate on plateau)
- Patience: 5 epochs before reduction
- Minimum learning rate: 1e-6
- **Secondary**: Cosine annealing with warm restarts
- **Warmup**: Linear warmup for first 5% of epochs

#### 6. Gradient Clipping and Stability
```python
def clip_gradients(self, max_norm: float = 1.0) -> float:
    """
    Apply gradient clipping for training stability
    
    Parameters:
    - max_norm: Maximum gradient norm threshold
    
    Returns:
    - grad_norm: Actual gradient norm before clipping
    """
```

**Gradient Management Features:**
- Gradient norm clipping with configurable threshold
- Gradient norm monitoring and logging
- Gradient explosion detection and handling
- Parameter update magnitude tracking
- Gradient flow analysis for debugging

#### 7. Comprehensive Training Metrics and Logging
```python
class TrainingMetrics:
    """Comprehensive metrics tracking and logging system"""
    
    def __init__(self, log_dir: str, experiment_name: str):
        """
        Initialize metrics tracking with file and tensorboard logging
        
        Parameters:
        - log_dir: Directory for logging outputs
        - experiment_name: Unique experiment identifier
        """
```

**Metrics Tracking Requirements:**
- **Training Metrics**: Loss, accuracy, learning rate, gradient norm
- **Validation Metrics**: Loss, AUC-ROC, AUC-PR, F1-score, precision, recall
- **System Metrics**: GPU memory usage, training speed, batch processing time
- **Model Metrics**: Parameter counts, activation statistics, weight norms
- **TensorBoard Integration**: Real-time visualization and experiment comparison
- **CSV Logging**: Structured data for analysis and reporting

### Training Configuration (from configs/model_config.yaml)
```yaml
lstm_model:
  training:
    batch_size: 32
    learning_rate: 0.001
    num_epochs: 100
    early_stopping_patience: 15
    weight_decay: 1e-4
    gradient_clip_norm: 1.0
    
  optimization:
    optimizer: "adam"
    beta1: 0.9
    beta2: 0.999
    epsilon: 1e-8
    amsgrad: false
    
  scheduler:
    type: "reduce_on_plateau"
    factor: 0.5
    patience: 5
    min_lr: 1e-6
    
  loss_function:
    type: "weighted_bce"
    pos_weight: 3.0
    label_smoothing: 0.0
    focal_loss: false
    
  early_stopping:
    patience: 15
    min_delta: 1e-4
    restore_best_weights: true
    monitor: "val_loss"
    
  logging:
    log_interval: 10  # Log every 10 batches
    save_interval: 5  # Save checkpoint every 5 epochs
    tensorboard_enabled: true
    csv_logging: true
```

### File Structure Extension
```
src/models/model_trainer.py
├── LSTMTrainer class (NEW)
├── EarlyStopping class (NEW)
├── TrainingMetrics class (NEW)
├── ModelCheckpoint class (NEW)
├── train_epoch method (NEW)
├── validate_epoch method (NEW)
├── train method (NEW)
└── utilities for training management (NEW)
```

### Training Loop Implementation
```python
def train(self, train_loader: DataLoader, val_loader: DataLoader) -> Dict:
    """
    Complete training loop with all advanced features
    
    Returns:
    - training_history: Dictionary containing all training metrics
    """
    
    # Training phases:
    # 1. Model initialization and device setup
    # 2. Optimizer and scheduler configuration
    # 3. Early stopping and checkpoint setup
    # 4. Training loop with validation
    # 5. Best model restoration and final evaluation
```

### Single Epoch Training
```python
def train_epoch(self, train_loader: DataLoader, epoch: int) -> Dict:
    """
    Single epoch training with comprehensive monitoring
    
    Features:
    - Batch-wise loss computation and backpropagation
    - Gradient clipping and optimization step
    - Real-time metrics calculation and logging
    - Progress tracking with tqdm
    - Memory usage monitoring
    """
```

### Validation Implementation
```python
def validate_epoch(self, val_loader: DataLoader, epoch: int) -> Dict:
    """
    Single epoch validation with detailed metrics
    
    Features:
    - Model evaluation mode with gradient disabled
    - Comprehensive metric calculation (AUC-ROC, AUC-PR, F1)
    - Prediction probability distribution analysis
    - Class-wise performance evaluation
    - Model interpretability metrics
    """
```

### Model Checkpointing
```python
class ModelCheckpoint:
    """Advanced model checkpointing with state management"""
    
    def save_checkpoint(self, model: nn.Module, optimizer: optim.Optimizer,
                       scheduler, epoch: int, metrics: Dict) -> None:
        """
        Save comprehensive model checkpoint
        
        Includes:
        - Model state dict and architecture config
        - Optimizer and scheduler states
        - Training metrics and epoch information
        - Random states for reproducibility
        """
```

### Training Speed Targets
- **Single Epoch**: < 2 minutes on GPU for 1000 sequences
- **Full Training**: < 30 minutes total (100 epochs with early stopping)
- **Memory Usage**: < 4GB GPU memory during training
- **Batch Processing**: < 200ms per batch (32 sequences)

### Monitoring and Logging Performance
- **Metrics Calculation**: < 50ms per epoch
- **TensorBoard Logging**: < 100ms per epoch
- **Checkpoint Saving**: < 5 seconds per checkpoint
- **Memory Profiling**: Minimal overhead (< 5% slowdown)

### Unit Tests (`tests/test_lstm_trainer.py`)
```python
class TestLSTMTrainer:
    def test_trainer_initialization(self):
        """Test trainer setup with various configurations"""
        
    def test_optimizer_configuration(self):
        """Test optimizer setup and parameter groups"""
        
    def test_loss_function_weighting(self):
        """Test weighted loss calculation and class balance"""
        
    def test_early_stopping_logic(self):
        """Test early stopping trigger conditions"""
        
    def test_learning_rate_scheduling(self):
        """Test LR scheduler behavior"""
        
    def test_gradient_clipping(self):
        """Test gradient clipping functionality"""
        
    def test_metrics_tracking(self):
        """Test comprehensive metrics calculation"""
        
    def test_checkpoint_saving_loading(self):
        """Test model checkpoint save/load functionality"""
        
    def test_gpu_cpu_compatibility(self):
        """Test training on both GPU and CPU"""
        
    def test_training_resumption(self):
        """Test training continuation from checkpoint"""
```

### Integration Tests
- End-to-end training pipeline with mock data
- GPU memory management and optimization
- TensorBoard logging and visualization
- Multi-epoch training stability
- Early stopping trigger validation
- Model convergence testing

### Core Implementation
1. `LSTMTrainer` class with complete training pipeline
2. `EarlyStopping` class with validation monitoring
3. `TrainingMetrics` class with comprehensive logging
4. `ModelCheckpoint` class with state management
5. Optimizer configuration and management utilities
6. Learning rate scheduling implementation
7. Gradient clipping and stability features

### Supporting Components
8. Comprehensive unit test suite (`tests/test_lstm_trainer.py`)
9. Integration tests with sequence dataset
10. Performance benchmarking utilities
11. TensorBoard logging setup and configuration
12. Training progress visualization tools
13. Model debugging and analysis utilities
14. Training pipeline documentation

### Functional Validation
- [ ] Training pipeline successfully trains LSTM model to convergence
- [ ] Early stopping triggers correctly based on validation metrics
- [ ] Weighted loss function properly handles class imbalance
- [ ] Learning rate scheduling improves training stability
- [ ] Gradient clipping prevents training instability
- [ ] Comprehensive metrics are accurately calculated and logged

### Performance Validation
- [ ] Training completes within 30 minutes on GPU
- [ ] Memory usage stays under 4GB GPU memory
- [ ] Single epoch processing time under 2 minutes
- [ ] TensorBoard logging works without performance degradation
- [ ] Checkpoint saving/loading operates efficiently

### Integration Validation
- [ ] Seamless integration with LSTM model architecture
- [ ] Proper interface with sequence dataset and data loaders
- [ ] Configuration system controls all training parameters
- [ ] GPU acceleration works correctly and improves speed
- [ ] Training can be resumed from checkpoints without issues

### Basic Training Setup
```python
# Initialize model and trainer
from src.models.lstm_model import SteelDefectLSTM
from src.models.model_trainer import LSTMTrainer
from src.data.sequence_dataset import create_data_loaders

# Load configuration
with open('configs/model_config.yaml', 'r') as f:
    config = yaml.safe_load(f)

# Initialize model
model = SteelDefectLSTM(
    input_size=config['lstm_model']['architecture']['input_size'],
    hidden_size=config['lstm_model']['architecture']['hidden_size'],
    num_layers=config['lstm_model']['architecture']['num_layers'],
    dropout=config['lstm_model']['architecture']['dropout']
)

# Initialize trainer
trainer = LSTMTrainer(model, config['lstm_model'])

# Train model
train_loader, val_loader = create_data_loaders(train_dataset, val_dataset, config)
training_history = trainer.train(train_loader, val_loader)
```

### Advanced Training with Custom Configuration
```python
# Custom training configuration
custom_config = {
    'training': {
        'learning_rate': 0.0005,
        'weight_decay': 5e-5,
        'gradient_clip_norm': 0.5
    },
    'early_stopping': {
        'patience': 20,
        'min_delta': 1e-5
    },
    'scheduler': {
        'type': 'cosine_annealing',
        'T_max': 50
    }
}

# Initialize with custom configuration
trainer = LSTMTrainer(model, custom_config, device='cuda')

# Train with experiment tracking
import wandb
wandb.init(project="steel-defect-lstm")
training_history = trainer.train(train_loader, val_loader)
wandb.log(training_history)
```

### Training Resumption from Checkpoint
```python
# Resume training from saved checkpoint
checkpoint_path = 'models/deep_learning/checkpoint_epoch_25.pth'
trainer.load_checkpoint(checkpoint_path)

# Continue training
remaining_history = trainer.train(
    train_loader, val_loader, 
    start_epoch=26, 
    max_epochs=100
)
```

### Required Packages
```python
torch>=1.13.0
numpy>=1.23.0
pyyaml>=6.0
tqdm>=4.64.0
tensorboard>=2.11.0
scikit-learn>=1.1.0  # For metrics calculation
matplotlib>=3.6.0    # For visualization
wandb>=0.13.0        # Optional for experiment tracking
```

### Development and Testing Tools
```python
pytest>=7.2.0
pytest-cov>=4.0.0
memory-profiler>=0.60.0
torch-tb-profiler>=0.4.0  # For advanced profiling
```

### Phase 1: Core Training Loop (Week 1)
- [ ] Basic `LSTMTrainer` class structure
- [ ] Adam optimizer configuration
- [ ] Simple training and validation loops
- [ ] Basic metrics calculation

### Phase 2: Advanced Features (Week 1-2)
- [ ] Early stopping implementation
- [ ] Learning rate scheduling
- [ ] Gradient clipping and stability
- [ ] Weighted loss function implementation

### Phase 3: Logging and Monitoring (Week 2)
- [ ] Comprehensive metrics tracking
- [ ] TensorBoard integration
- [ ] Model checkpointing system
- [ ] Training progress visualization

### Phase 4: Testing and Optimization (Week 2)
- [ ] Complete unit test suite
- [ ] Integration testing
- [ ] Performance optimization
- [ ] GPU acceleration validation

### Training Performance
- [ ] LSTM model achieves AUC-ROC > 0.88 on validation set
- [ ] Training converges within 100 epochs (typically 60-80 epochs)
- [ ] Early stopping triggers appropriately (no overfitting)
- [ ] Training time under 30 minutes on GPU
- [ ] Stable training without gradient explosion or vanishing

### System Performance
- [ ] Memory usage under 4GB GPU memory
- [ ] Efficient checkpoint saving and loading
- [ ] TensorBoard logging without performance impact
- [ ] Robust error handling and recovery
- [ ] Cross-platform compatibility (Windows, Linux, macOS)

This implementation forms the backbone of the LSTM model training system, providing enterprise-grade training capabilities with comprehensive monitoring, optimization, and debugging features for steel casting defect prediction.
